crates/gossip/src/storage_arc.rs

```diff
@@ -0,0 +1,290 @@
+/// in the snapshot. Such sectors are considered synced. If any synced sectors can be added to the
+/// start or end of the current storage arc for a local agent, it will do so. Any synced sectors
+/// that cannot be added to the storage arc to produce a larger continuous arc are ignored.
+pub(crate) fn update_storage_arcs(
+    snapshot: &DhtSnapshot,
+    local_agents: Vec<DynLocalAgent>,
+    common_arc_set: ArcSet,
+) -> K2Result<()> {
+    let mismatched_sectors = match snapshot {
+        DhtSnapshot::Minimal { .. } => HashSet::with_capacity(0),
+        DhtSnapshot::RingSectorDetails {
+            ring_sector_hashes, ..
+        } => {
+            // These sectors didn't match, so we can't include them in our storage arc.
+            ring_sector_hashes
+                .values()
+                .flat_map(|v| v.keys())
+                .copied()
+                .collect::<HashSet<_>>()
+        }
+        _ => {
+            tracing::info!("Unable to update storage arc with a non-ring sector details snapshot");
+            return Ok(());
+        }
+    };
+
+    // The difference between the common arc set used to construct the DhtSnapshot and the
+    // mismatched sectors is the set of sectors that we can use to update our storage arc.
+    let synced_arcs = common_arc_set
+        .without_sector_indices(mismatched_sectors)
+        .as_arcs();
+
+    for local_agent in local_agents {
+        let target_storage_arc = local_agent.get_tgt_storage_arc();
+        let current_storage_arc = local_agent.get_cur_storage_arc();
+
+        if target_storage_arc == current_storage_arc {
+            // No adjustment needed, already at the target.
+            continue;
+        }
+
+        let target_set = ArcSet::new(vec![target_storage_arc])?;
+
+        // Add our current storage arc to the set of sectors covered by synced arcs.
+        let mut synced_arc_set = synced_arcs.clone();
+        synced_arc_set.push(current_storage_arc);
+
+        // Then intersect the result with our target arc. This prevents the storage arc from
+        // growing outside the target arc. That might happen when we have multiple local agents
+        // with different target arcs.
+        let new_arcs = ArcSet::new(synced_arc_set)?
+            .intersection(&target_set)
+            .as_arcs();
+
+        if current_storage_arc == DhtArc::Empty {
+            // When our current storage arc is empty, we're free to pick a new one that is
+            // contained within the target arc. Pick the largest one.
+            if let Some(new_arc) =
+                new_arcs.into_iter().max_by_key(|arc| arc.len())
+            {
+                local_agent.set_cur_storage_arc(new_arc);
+            }
+        } else {
+            // At this point, we have added any synced sectors to our current storage arc and ignored
+            // and synced sectors that aren't in our target arc. We can find the largest overlapping
```

Comment:
not ignored and synced, I believe?
```suggestion
            // sectors that aren't in our target arc. We can find the largest overlapping
```

Reply:
Correct, that's redundant - I updated this locally